File: Difficulty: Medium/Number of pairs/number-of-pairs.py
from bisect import bisect_right

class Solution:
    def countPairs(self, arr, brr):
        # Count occurrences of 1, 2, 3, and 4 in brr because they are special cases
        brr.sort()
        count1 = brr.count(1)
        count2 = brr.count(2)
        count3 = brr.count(3)
        count4 = brr.count(4)
        
        total_pairs = 0
        
        for x in arr:
            if x == 1:
                # If x is 1, skip because 1^y < y^1 for all y > 1
                continue
            elif x == 2:
                # For x = 2, skip (2, 3) and (2, 4) as they don't satisfy the condition
                # Count all brr[j] > 2
                pos = bisect_right(brr, 2)
                total_pairs += len(brr) - pos
                # Add pairs with y = 1
                total_pairs += count1
                # Subtract pairs with y = 3 and y = 4
                total_pairs -= count3 + count4
            elif x == 3:
                # For x = 3, skip (3, 2) as it doesn't satisfy the condition
                # Count all brr[j] > 3
                pos = bisect_right(brr, 3)
                total_pairs += len(brr) - pos
                # Add pairs with y = 1
                total_pairs += count1
                # Add pairs with y = 2
                total_pairs += count2
            else:
                # For all other x > 3, count all brr[j] > x
                pos = bisect_right(brr, x)
                total_pairs += len(brr) - pos
                # Add pairs with y = 1
                total_pairs += count1
        
        return total_pairs


#{ 
 # Driver Code Starts
#Initial Template for Python 3

#Initial Template for Python 3

import atexit
import io
import sys
import bisect

_INPUT_LINES = sys.stdin.read().splitlines()
input = iter(_INPUT_LINES).__next__
_OUTPUT_BUFFER = io.StringIO()
sys.stdout = _OUTPUT_BUFFER


@atexit.register
def write():
    sys.__stdout__.write(_OUTPUT_BUFFER.getvalue())


if __name__ == '__main__':
    t = int(input())
    for i in range(t):
        # M, N = map(int, input().strip().split())
        a = list(map(int, input().strip().split()))
        b = list(map(int, input().strip().split()))
        ob = Solution()
        print(ob.countPairs(a, b))
        #code here

# } Driver Code Ends

File: Difficulty: Medium/Number of pairs/number-of-pairs.cpp
//{ Driver Code Starts
// Initial template for C++

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function template for C++

class Solution {
  public:
    long long countPairs(vector<int>& arr1, vector<int>& arr2) {
        sort(arr2.begin(), arr2.end());
        
        // Precompute counts for special values in arr2
        int count1 = 0, count2 = 0, count3 = 0, count4 = 0;
        for (int y : arr2) {
            if (y == 1) count1++;
            else if (y == 2) count2++;
            else if (y == 3) count3++;
            else if (y == 4) count4++;
        }
        
        long long result = 0;
        for (int x : arr1) {
            if (x == 1) {
                continue; // (1, y) does not satisfy x^y > y^x unless y == 1
            }

            // Binary search to count valid pairs
            int pos = upper_bound(arr2.begin(), arr2.end(), x) - arr2.begin();
            result += arr2.size() - pos;
            
            // Special cases:
            if (x == 2) {
                result -= count4 + count3; // For (2,3), (2,4) -> 2^y <= y^2
                result += count1;          // For (2,1) -> 2^1 > 1^2
            } else if (x == 3) {
                result += count2;          // For (3,2) -> 3^2 > 2^3
                result += count1;          // For (3,1) -> 3^1 > 1^3
            } else {
                result += count1;          // For (x,1) -> x^1 > 1^x
            }
        }

        return result;
    }
};

//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    cin.ignore(); // Ignore the newline character after t
    while (t--) {
        vector<int> arr1, arr2;
        string input1, input2;

        getline(cin, input1); // Read the entire line for the first array elements
        stringstream ss1(input1);
        int number;
        while (ss1 >> number) {
            arr1.push_back(number);
        }

        getline(cin, input2); // Read the entire line for the second array elements
        stringstream ss2(input2);
        while (ss2 >> number) {
            arr2.push_back(number);
        }

        Solution ob;
        long long ans = ob.countPairs(arr1, arr2);
        cout << ans << "\n~\n";
    }
    return 0;
}
// } Driver Code Ends